### ISSUE TITLE
Toil can't saturate Slurm with jobs because it spends all its time polling jobs already submitted

### ISSUE THREAD

**Issue by edraizen** [DataBiosphere/toil#2323]
I am allowed to submit/run 10,000 jobs at once using my institutes slurm hpc. However, toil only submits about 300 at a time. 

This is how I initialize it:
```shell
export TOIL_SLURM_ARGS="-t 10:00:00 --partition=standard"
python test.py --defaultCores=4 --defaultMemory=12000M --maxLocalJobs=10000 --batchSystem=slurm default
```
where test.py starts one job, and that job has ~20,000 child jobs and then ~20,000 follow on jobs. Is there another way to start these jobs? Is setting maxLocalJobs right?

Thanks,
Eli

**Comment by DailyDreaming** [comment]
Hmmm... could you try setting "--targetTime=0"?  It's something of an obscure option, but it should launch jobs more aggressively.

**Comment by jsmedmar** [comment]
I'm having the same issue on LSF, tried `--targetTime=0` but complained about requiring a positive integer. Tried with `--targetTime=1`, but toil doesn't submit more than 100 jobs at the time. @DailyDreaming should I try changing this validation and allow 0?

**Comment by jsmedmar** [comment]
I managed to set it to 0, but nothing changed. Actually `targetTime` doesn't seem to be used at all in the codebase , see this [search](https://github.com/DataBiosphere/toil/search?q=targetTime&unscoped_q=targetTime).

**Comment by joelarmstrong** [comment]
It is used (github only shows the top 2 hits per file, so those search results only catch the docstring mentions), but only in the autoscaler code, so it shouldn't have any effect when running on an institutional cluster.

**Comment by jsmedmar** [comment]
Ah I see, this function is not used at all for sure: `getWaitDuration`.

**Comment by jsmedmar** [comment]
I couldn't figure out while toil seems to throttle the job submission... I think this shouldn't be the case. as @edraizen on should be able to submit thousands at the time. I'll look more into it next week.

**Comment by jsmedmar** [comment]
I think it has to do with this [line](https://github.com/DataBiosphere/toil/blob/master/src/toil/batchSystems/abstractGridEngineBatchSystem.py#L112). It makes sense to limit the number of jobs to the total number of cores of the machine when using `singleMachine` as the batch system. However, clusters should be able to run much more. I might be wrong, specially if `maxCores` points to the total number of cores available in the cluster. Although max number of cores in cluster is actually saved in `maxCPU`.

**Comment by DailyDreaming** [comment]
@jsmedmar I think #2308 is related to that same line.  Looks like it might be a simple fix (thanks for the investigatory work!).  Currently having @bencvdb write up a fix that hopefully addresses both issues.

**Comment by DailyDreaming** [comment]
#2347 was merged.  Closing this.

Please feel free to reopen if #2347 did not in fact solve the problem.

**Comment by jsmedmar** [comment]
Hey @DailyDreaming, my cluster is on maintenance! Will check as soon as its up and running again.

**Comment by jsmedmar** [comment]
Want to let you know that #2347 did fix the problem! Nice work @bencvdb, thank you very much. For anyone landing here, I found that in order to get thousands of jobs fast its important to use these flags in combination:

- `--maxLocalJobs`: Set it to a large number, say `2000` or whatever your cluster can support.
- `--statePollingWait`: Use a larger delay for polling wait, this means check on jobs less often. Default is 1 second, that means that the toil worker would be querying the `batchSystem` to often. Do something like five minutes `5*60`. Toil will spend 5 mins submitting jobs instead of submitting **and** checking on jobs.

My last suggestion to toil developers would be to default `maxLocalJobs` to [`maxCpu`](https://github.com/DataBiosphere/toil/blob/master/src/toil/batchSystems/abstractGridEngineBatchSystem.py#L291). Although further consideration may be needed for that.

**Comment by jasonsydes** [comment]
Hello @bencvdb, @jsmedmar, @edraizen, and @joelarmstrong, and others,

I'm experiencing the same problem described in this issue.  

I would be curious to hear your experience.  @jsmedmar, when you say that "#2347 did fix the problem", could you elaborate?  Are you seeing thousands of jobs (RUNNING or PENDING) in your Slurm queue?

My experiences is as follows.  I too am allowed to submit a ridiculous number of jobs to our cluster via Slurm (and I have confirmed this by submitting thousands independently of / unrelated to Toil/Cactus).  When I launch cactus on our test alignment with 4 species, I watch the number of RUNNING and PENDING jobs in the Slurm queue, and I also watch the number of Issued jobs in the cactus.log.  The number of RUNNING jobs will vary between 0 and ~400 (it struggles to get to 400, usually stays around 200), the number of PENDING jobs is usually 0, but occasionally bounces to 5 or 10 or even 20.  However the number of outstanding Issued jobs, is in the thousands. 

For example, I can see that last night while I slept, there were 8240 'RunBlast' jobs issued in under 7 minutes (between 00:41:38am and 00:48:05am).  

I've tried many invocations of cactus per some of the comments in Toil / Cactus github issues I've seen (e.g. this one, plus #2347, #928 and #2509), but they all seem to behave the same.  Right now, this is my current invocation: 
```
cactus --batchSystem slurm --maxLocalJobs 4000 --servicePollingInterval 300 \
      --disableCaching --logFile cactus.log --binariesMode local \
      --stats jobStore seqFile.txt out.hal
```
(I'm attaching my full script here as well in case its useful: [cactus.sh.txt](https://github.com/DataBiosphere/toil/files/3118296/cactus.sh.txt)).


I can see the number of jobs in the Slurm queue jump from 0 to 100 or 200 rather rapidly (in about one minute or so).  But I've never seen it jump anywhere close to 4,000.

My concern is that Toil seems to be "holding back" and not submitting already "Issued" jobs to Slurm.  Most of the jobs in the Slurm queue are in the "RUNNING" state.  Rarely do I see jobs sitting in the PENDING state on the queue.  

It would be preferable if Toil were able to move many more of its jobs from what appears to be its internal queue out into Slurm's queue, where a) they could "start the clock" on getting allocated resources from Slurm (i.e. sit in the Slurm queue waiting in the PENDING state), and b) they could actually run (I'm positive that our cluster will allow me to have many more jobs in the RUNNING state than Toil/Cactus is currently throwing at it).

I've written a script to simultaneously watch the Slurm queue and the cactus log every 5 seconds (see the PS below for Field Descriptions).  Here you can see there are 191 jobs Running on Slurm (they happen to be LastzRepeatMaskJob jobs), there are none Pending in the Slurm queue, and there are 1414 - 241 = 1,173 jobs sitting in Toil's "internal queue":
```
Running: 191, Pending: 0, Issued L: 1414, Ended L: 241, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:43:09 PDT 2019
```
That's just one line, there are many more like it:
```
Running: 131, Pending: 0, Issued L: 1414, Ended L: 160, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:39:45 PDT 2019
Running: 128, Pending: 0, Issued L: 1414, Ended L: 160, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:39:50 PDT 2019
Running: 133, Pending: 1, Issued L: 1414, Ended L: 161, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:39:55 PDT 2019
Running: 132, Pending: 0, Issued L: 1414, Ended L: 168, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:40:00 PDT 2019
Running: 129, Pending: 0, Issued L: 1414, Ended L: 169, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:40:05 PDT 2019
Running: 132, Pending: 0, Issued L: 1414, Ended L: 178, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:40:10 PDT 2019
Running: 128, Pending: 0, Issued L: 1414, Ended L: 180, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:40:15 PDT 2019
Running: 125, Pending: 0, Issued L: 1414, Ended L: 180, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:40:21 PDT 2019
Running: 130, Pending: 0, Issued L: 1414, Ended L: 188, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:40:26 PDT 2019
Running: 127, Pending: 0, Issued L: 1414, Ended L: 188, Issued B: 0, Ended B: 0. -- Thu Apr 25 10:40:31 PDT 2019
```
As you can see, Toil is just trickling the jobs from its internal queue to the Slurm queue.

Here attached is the full log of that script for the first 16 minutes of the cactus pipeline, along with the script itself and a description of the fields: 
[monitor_slurm_queue_and_cactus_log.txt](https://github.com/DataBiosphere/toil/files/3118324/monitor_slurm_queue_and_cactus_log.txt)

I'm using the LastzRepeatMaskJob as an example here, because it's convenient (it happens early in the pipeline).  The situation is worse for the thousands of RunBlast jobs that are Issued later.  I'm running my script to log that and will attach it later when it is available.

Am I doing something wrong?  Could I be doing anything differently?

I would reopen this ticket if I could, but I think I would need to be the original author or an administrator.  I wasn't sure whether to post here or start a new issue.  As far as I can tell, in my environment, the issue of "slurm max jobs not met" still exists.  While cactus finishes in a reasonable amount of time on this 4-species alignment, I am concerned that this problem will significantly increase total execution time when we start adding more species.

Thank you for your assistance!
Jason

PS Field descriptions of my script's output:
    Running: Number of RUNNING jobs in the Slurm queue (matching "toil_job_\*").
    Pending: Number of PENDING jobs in the Slurm queue (matching "toil_job_\*").
    Issued L: Number of 'Issued' 'LastzRepeatMaskJob' jobs in cactus.log.
    Ended L:  Number of 'Job ended' 'LastzRepeatMaskJob' jobs in cactus.log.
    Issued B: Number of 'Issued' 'RunBlast' jobs in cactus.log.
    Ended B:  Number of 'Job ended' 'RunBlast' jobs in cactus.log.

**Comment by jasonsydes** [comment]
As promised, here is the output of my summary script watching a 4-species cactus run well into the first set of RunBlast jobs: 

[monitor_slurm_queue_and_cactus_log.with_some_RunBlast_jobs.txt](https://github.com/DataBiosphere/toil/files/3118662/monitor_slurm_queue_and_cactus_log.with_some_RunBlast_jobs.txt)

Here is the cactus.log from which the output above was generated:

[cactus.log](https://github.com/DataBiosphere/toil/files/3118642/cactus.log)

In my scripts output, you can see that starting at 12:00:17, cactus generates 8,240 RunBlast jobs in less than 7 minutes. But in the 45 minutes following that issuance of jobs, there are never more than 265 jobs RUNNING in the Slurm queue, and no more than 49 jobs PENDING in the Slurm queue (and the vast majority of that time there are 0 jobs PENDING).

Why is Toil hanging onto those jobs instead of submitting them to Slurm?

**Comment by joelarmstrong** [comment]
I've reopened the issue. Yeah, that seems very, very strange. I'm not totally sure what the current status of toil's Slurm support is. It looks like you've been very thorough (thank you!) in trying to work around any existing bugs I know of (the default setting / mislabeling of `--maxLocalJobs` being the biggest offender).

I don't have access to a slurm cluster to check, but according to my (possibly outdated) understanding of the slurm code this *should* be issuing at least 4,000 jobs at once, but clearly it's not. I'll try to figure out where things are going wrong. Just thinking out loud, maybe there's only at most one job issued per "poll" of the scheduler state?

One thing you might try -- on the off-chance that someone fixed the `--maxLocalJobs`/`--maxCores` behavior swap on gridengine/slurm/LSF, is to try using `--maxCores 4000`. It'd be a long shot, though, I don't know if that code's been changed in a while.

**Comment by jasonsydes** [comment]
Thanks for getting back to this so quickly Joel, and thank you for your upcoming efforts to hunt down the problem!

I tried --maxCores 4000, but it's the same result:

[mon.v1.txt](https://github.com/DataBiosphere/toil/files/3119096/mon.v1.txt)
[cactus.log](https://github.com/DataBiosphere/toil/files/3119097/cactus.log)
[20190425-with-maxCores.sb.txt](https://github.com/DataBiosphere/toil/files/3119099/20190425-with-maxCores.sb.txt)

I'm not entirely sure what you mean by "*maybe there's only at most one job issued per "poll" of the scheduler state*", maybe that's something internal to Toil?

I'm happy to continue running tests on our slurm cluster to assist.  Alternatively, if it would help, we might be able to provide you with a debugging account on our cluster; I'll look into that, see if that's possible.

**Comment by jsmedmar** [comment]
disclaimer - haven't read full thread (will do later this week). Nevertheless you need to give toil time to query the batch system (btw we use `LSF`). I suggest to try `--statePollingWait 300` and see if you get more jobs. Toil will try to submit jobs for 5 minutes before checking on jobs again. Toil (at least the default LSF implementation) runs `bjobs` for every job every `statePollingWait` seconds. Running `bjobs` is super slow, hence toil ends up using all the `statePollingWait` time checking on jobs and never manages to submit new jobs... 

We addressed this issue in our `CustomLSF` batch system by making one `bjobs` call to get the status of all current running jobs. Meaning if there are 10K jobs running, we call `bjobs` once instead of 10 thousand times:

https://github.com/papaemmelab/toil_container/blob/2afa80080fc83a34491cd80efe05159ddc386729/toil_container/lsf.py#L107

Again to rule out this possibility, use `--statePollingWait 300` and see if you get more jobs in the queue. I'll read your thorough question later and re-update this msg if needed.

**Comment by jasonsydes** [comment]
Hi @jsmedmar!  

Oh goodness, I had seen your comment in #2509 about setting --statePollingWait to 300, but along the way apparently I confused that option with --servicePollingInterval and had used that instead (see in my script above: --servicePollingInterval 300).  Oops.

I will try this shortly and report back.  Thank you for your feedback!
Jason

**Comment by jasonsydes** [comment]
Hi @jsmedmar and @joelarmstrong,

I'm pretty sure that @jsmedmar's suggestion did the trick.  I first tried --statePollingWait 300, but (maybe??? I haven't looked deeply) that seemed to pause too long after just the first job?  (I have not looked deep, so grain of salt...).  I dropped it to --statePollingWait to 30, and that seems to have worked!!

I will report more shortly, but:
```
Running: 4, Pending: 0, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:49:30 PDT 2019
Running: 4, Pending: 1, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:49:35 PDT 2019
Running: 13, Pending: 0, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:49:40 PDT 2019
Running: 12, Pending: 0, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:49:45 PDT 2019
Running: 7, Pending: 5, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:49:50 PDT 2019
Running: 17, Pending: 0, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:49:55 PDT 2019
Running: 13, Pending: 0, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:02 PDT 2019
Running: 6, Pending: 0, Issued L: 0, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:07 PDT 2019
Running: 38, Pending: 0, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:12 PDT 2019
Running: 66, Pending: 2, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:17 PDT 2019
Running: 118, Pending: 0, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:24 PDT 2019
Running: 145, Pending: 0, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:32 PDT 2019
Running: 183, Pending: 0, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:38 PDT 2019
Running: 182, Pending: 0, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:44 PDT 2019
Running: 180, Pending: 0, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:49 PDT 2019
Running: 176, Pending: 16, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:54 PDT 2019
Running: 193, Pending: 1, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:50:59 PDT 2019
Running: 218, Pending: 14, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:05 PDT 2019
Running: 233, Pending: 38, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:10 PDT 2019
Running: 228, Pending: 26, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:15 PDT 2019
Running: 220, Pending: 28, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:20 PDT 2019
Running: 231, Pending: 26, Issued L: 242, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:25 PDT 2019
Running: 223, Pending: 26, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:31 PDT 2019
Running: 222, Pending: 26, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:36 PDT 2019
Running: 230, Pending: 26, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:41 PDT 2019
Running: 234, Pending: 26, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:47 PDT 2019
Running: 230, Pending: 27, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:52 PDT 2019
Running: 228, Pending: 46, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:51:57 PDT 2019
Running: 230, Pending: 110, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:52:02 PDT 2019
Running: 218, Pending: 205, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:52:07 PDT 2019
Running: 218, Pending: 303, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:52:12 PDT 2019
Running: 218, Pending: 392, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:52:18 PDT 2019
Running: 278, Pending: 344, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:52:23 PDT 2019
Running: 228, Pending: 358, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:52:32 PDT 2019
Running: 218, Pending: 362, Issued L: 558, Ended L: 0, Issued B: 0, Ended B: 0. -- Thu Apr 25 19:52:37 PDT 2019
```
Thank you!!!
Jason

**Comment by jasonsydes** [comment]
This may not be the panacea that I was hoping it would be.  I'm looking at some numbers now.  But the idea is, by applying --statePollingWait 300, you are causing very short jobs that are run serially to take much longer to launch.  

So, if you run 6 jobs that run serially and that take on average 22 seconds to complete (129s total), then by applying --statePollingWait 300 you'll now cause those same 6 jobs to complete in 5*300s =
1800 seconds.  Here is that exact example (this is output from a cactus.log watcher script I wrote) from the first 6 completed jobs:
```
# without --statePollingWait 300

∴ head -500 watch.issued.* |grep Ended 
2019-04-24 23:31:03 Job Ended (0h 0m 9s or 9s): RunCactusPreprocessorThenProgressiveDown (1) z/V/jobnLircL (1) 20190424-4way-mlj3000 (toil_job_0 and cores: 1, disk: 2.0 G, and memory: 6.5 G)
2019-04-24 23:31:24 Job Ended (0h 0m 20s or 20s): logAssemblyStats (1) A/3/job41kzUe (1) 20190424-4way-mlj3000 (toil_job_1 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-24 23:31:38 Job Ended (0h 0m 35s or 35s): logAssemblyStats (2) L/W/jobkVQP3q (1) 20190424-4way-mlj3000 (toil_job_2 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-24 23:31:38 Job Ended (0h 0m 35s or 35s): logAssemblyStats (3) 0/q/jobsoMyNI (1) 20190424-4way-mlj3000 (toil_job_3 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-24 23:31:29 Job Ended (0h 0m 25s or 25s): logAssemblyStats (4) j/q/jobXh4G14 (1) 20190424-4way-mlj3000 (toil_job_4 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-24 23:31:09 Job Ended (0h 0m 5s or 5s): CactusPreprocessor (1) j/2/job_g9Hj0 (1) 20190424-4way-mlj3000 (toil_job_5 and cores: 1, disk: 4.1 G, and memory: 2.0 G)

# with --statePollingWait 300

∴ cat watch.issued.20190426.091008.log | grep Ended
2019-04-26 09:15:17 Job Ended (0h 5m 1s or 301s): RunCactusPreprocessorThenProgressiveDown (1) o/O/jobofgn5p (1) 20190426-4way-slow-queuing---with--statePollingWait-300 (toil_job_0 and cores: 1, disk: 2.0 G, and memory: 6.5 G)
2019-04-26 09:20:17 Job Ended (0h 5m 0s or 300s): logAssemblyStats (1) n/o/jobgk2Ldh (1) 20190426-4way-slow-queuing---with--statePollingWait-300 (toil_job_1 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-26 09:20:17 Job Ended (0h 5m 0s or 300s): logAssemblyStats (2) X/6/jobLadTWl (1) 20190426-4way-slow-queuing---with--statePollingWait-300 (toil_job_2 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-26 09:20:17 Job Ended (0h 5m 0s or 300s): logAssemblyStats (3) f/7/jobO_1TUR (1) 20190426-4way-slow-queuing---with--statePollingWait-300 (toil_job_3 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-26 09:20:17 Job Ended (0h 5m 0s or 300s): logAssemblyStats (4) n/Q/jobAouNLw (1) 20190426-4way-slow-queuing---with--statePollingWait-300 (toil_job_4 and cores: 1, disk: 2.0 G, and memory: 2.0 G)
2019-04-26 09:20:17 Job Ended (0h 5m 0s or 300s): CactusPreprocessor (1) j/1/jobY0TX_6 (1) 20190426-4way-slow-queuing---with--statePollingWait-300 (toil_job_5 and cores: 1, disk: 4.1 G, and memory: 2.0 G)
```
This may add significant amount of time to total cactus pipeline execution time.  I'm trying to get a better sense of how much by looking at the numbers now.

**Comment by jasonsydes** [comment]
Hi again @jsmedmar and @joelarmstrong,

Well, with the information I have (which isn't much), I don't think I can come to a conclusion about the impact of using --statePollingWait.

To come a strong conclusion about the impact of using versus not using --statePollingWait 300 (or --statePollingWait 120 or --statePollingWait 30), I would need either a) a deeper understanding of the cactus pipeline, i.e. how many jobs are being run serially vs in parallel, and/or b) to have run the pipeline under many more conditions (unfortunately I don't have unlimited computational resources, so I likely won't be able to do this).

For now, I'll just make an educated guess. 

I'm not sure about --statePollingWait 300.  Under that condition, simply starting up the pipeline feels like it takes too long (after 10 minutes it's only to toil_job_9; in contrast, with the default --statePollingWait, the LastzRepeatMaskJob jobs are already running at 10 minutes in).  I'll admit I stopped it after 10 minutes, so it might be worthwhile to go back to this and try running it again for a longer amount of time.  

In contrast, --statePollingWait 30 starts more quickly, but it still seems to struggle to "fill the slurm queue".  It maxes out at 903 jobs in the Slurm queue (sum of RUNNING and PENDING).

With --statePollingWait 120, toil gets me to a maximum of 1,829 jobs in the Slurm queue (sum of RUNNING and PENDING) (but generally lower than 1,829 jobs). That's not getting up to my test setting of --maxLocalJobs 3500 (i.e. not saturating), but I'm just one user on our Slurm cluster, and 1,829 seems like a reasonable effective maximum.  

It obviously would be nice if Toil were able to simply submit the maxLocalJobs in one fell swoop, without having to increase the polling time.  I'm not nearly as familiar with the job submission code as you two obviously are, so I don't know how hard it would be to make that a reality.

Thank you again for your assistance!  It looks like --statePollingWait 120 will be an improvement!

**Comment by adamnovak** [comment]
I've retitled the issue to reflect what it sounds like the problem is here; we spend too much time polling the jobs already submitted, and we then can't submit new jobs fast enough, even though they are available in the workflow.

**Comment by adamnovak** [comment]
We should redesign the Slurm batch system so that it can ask Slurm just for the finished jobs, so that we don't need to do O(n) work to get one finished job when n jobs are running, and so it won't be O(n^2) work to find n finished jobs.

**Comment by adamnovak** [comment]
I am going to close this in favor of #4431 since Jira doesn't seem to want to sync that this is re-opened, because it was synced to the old Jira before, I think.
